File: go-spider/internal/security/http_transport.go
package security

import (
	"bytes"
	"fmt"
	"io"
	"net"
	"strconv"
	"strings"
	"time"
)

// SecureHTTPError is a stable secure transport failure.
type SecureHTTPError struct {
	ReasonCode string
}

func (e *SecureHTTPError) Error() string { return e.ReasonCode }

func secureHTTPFail(reason string) error { return &SecureHTTPError{ReasonCode: reason} }

// SecureHTTPRequest is an immutable request model.
type SecureHTTPRequest struct {
	Profile string
	Method  string
	URL     string
	Headers []Header
	Body    []byte
}

// Header is a single HTTP header pair.
type Header struct {
	Name  string
	Value string
}

// SecureHTTPResponse is the bounded response result.
type SecureHTTPResponse struct {
	StatusCode   int
	FinalURL     string
	Headers      []Header
	Body         []byte
	RedirectHops int
}

func invalidHeaderValue(value string) bool {
	return strings.ContainsAny(value, "\r\n\x00")
}

func isHTTPToken(value string) bool {
	if value == "" {
		return false
	}
	for _, r := range value {
		if r > 127 {
			return false
		}
		switch {
		case r >= '0' && r <= '9':
		case r >= 'A' && r <= 'Z':
		case r >= 'a' && r <= 'z':
		case strings.ContainsRune("!#$%&'*+-.^_`|~", r):
		default:
			return false
		}
	}
	return true
}

func trimOWS(value []byte) []byte {
	return bytes.Trim(value, " \t")
}

// ValidateSecureRequest validates method, body, and headers.
func ValidateSecureRequest(req SecureHTTPRequest) error {
	switch req.Method {
	case "GET", "HEAD", "POST":
	default:
		return secureHTTPFail("method_not_allowed")
	}
	if req.Method == "GET" || req.Method == "HEAD" {
		if len(req.Body) != 0 {
			return secureHTTPFail("method_body_not_allowed")
		}
	}
	if err := CheckRequestBodySize(len(req.Body)); err != nil {
		return err
	}
	for _, header := range req.Headers {
		if invalidHeaderValue(header.Name) || invalidHeaderValue(header.Value) {
			return secureHTTPFail("invalid_header")
		}
		if !isHTTPToken(header.Name) {
			return secureHTTPFail("invalid_header")
		}
		switch strings.ToLower(header.Name) {
		case "host", "transfer-encoding", "proxy-authorization", "connection", "content-length":
			return secureHTTPFail("forbidden_header")
		}
	}
	return nil
}

// BuildRequestBytes builds the raw HTTP/1.1 request.
func BuildRequestBytes(req SecureHTTPRequest, host string) ([]byte, error) {
	if err := ValidateSecureRequest(req); err != nil {
		return nil, err
	}
	idx := strings.Index(req.URL, "://")
	rest := req.URL
	if idx >= 0 {
		rest = req.URL[idx+3:]
	}
	slash := strings.IndexByte(rest, '/')
	path := "/"
	if slash >= 0 {
		path = rest[slash:]
	}
	if path == "" {
		path = "/"
	}
	var buf bytes.Buffer
	fmt.Fprintf(&buf, "%s %s HTTP/1.1\r\n", req.Method, path)
	fmt.Fprintf(&buf, "Host: %s\r\n", host)
	sentEncoding := false
	for _, header := range req.Headers {
		if strings.EqualFold(header.Name, "Accept-Encoding") {
			sentEncoding = true
		}
		fmt.Fprintf(&buf, "%s: %s\r\n", header.Name, header.Value)
	}
	if !sentEncoding {
		buf.WriteString("Accept-Encoding: identity\r\n")
	}
	if req.Method == "POST" {
		fmt.Fprintf(&buf, "Content-Length: %d\r\n", len(req.Body))
	}
	buf.WriteString("Connection: close\r\n\r\n")
	buf.Write(req.Body)
	return buf.Bytes(), nil
}

func parseResponseHead(head []byte) (int, []Header, error) {
	lines := bytes.Split(head, []byte("\r\n"))
	firstLine := lines[0]
	for _, b := range firstLine {
		if b < 0x20 && b != 0x09 && b != 0x20 {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		if b == 0x7f {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
	}
	statusParts := bytes.SplitN(firstLine, []byte(" "), 3)
	if len(statusParts) < 2 {
		return 0, nil, secureHTTPFail("invalid_http_response")
	}
	version := string(statusParts[0])
	if version != "HTTP/1.0" && version != "HTTP/1.1" {
		return 0, nil, secureHTTPFail("invalid_http_response")
	}
	statusText := statusParts[1]
	if len(statusText) != 3 {
		return 0, nil, secureHTTPFail("invalid_http_response")
	}
	for _, b := range statusText {
		if b < '0' || b > '9' {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
	}
	status, err := strconv.Atoi(string(statusText))
	if err != nil {
		return 0, nil, secureHTTPFail("invalid_http_response")
	}
	headers := make([]Header, 0, len(lines)-1)
	for _, line := range lines[1:] {
		if len(line) == 0 {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		if line[0] == ' ' || line[0] == '\t' {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		if bytes.ContainsAny(line, "\r\n\x00") {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		name, value, ok := bytes.Cut(line, []byte(":"))
		if !ok || len(name) == 0 {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		if !isHTTPToken(string(name)) {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		valueText := string(trimOWS(value))
		if strings.ContainsAny(valueText, "\r\n\x00") {
			return 0, nil, secureHTTPFail("invalid_http_response")
		}
		headers = append(headers, Header{Name: string(name), Value: valueText})
	}
	return status, headers, nil
}

func readResponseHeads(conn net.Conn, timeoutMS int) (int, []Header, []byte, error) {
	_ = conn.SetReadDeadline(time.Now().Add(time.Duration(timeoutMS) * time.Millisecond))
	pending := make([]byte, 0, 4096)
	tmp := make([]byte, 4096)
	used := 0
	for {
		if idx := bytes.Index(pending, []byte("\r\n\r\n")); idx >= 0 {
			headEnd := idx + 4
			total := used + headEnd
			if total > NewTransportBudget().ResponseHeadersBytes() {
				return 0, nil, nil, secureHTTPFail("response_headers_too_large")
			}
			status, headers, err := parseResponseHead(pending[:idx])
			if err != nil {
				return 0, nil, nil, err
			}
			pending = pending[headEnd:]
			used = total
			if status == 101 {
				return 0, nil, nil, secureHTTPFail("protocol_upgrade_not_allowed")
			}
			if status >= 100 && status <= 199 {
				continue
			}
			if status < 200 || status > 599 {
				return 0, nil, nil, secureHTTPFail("invalid_http_response")
			}
			return status, headers, pending, nil
		}
		if used+len(pending) >= NewTransportBudget().ResponseHeadersBytes() {
			return 0, nil, nil, secureHTTPFail("response_headers_too_large")
		}
		n, err := conn.Read(tmp)
		if n > 0 {
			pending = append(pending, tmp[:n]...)
		}
		if err != nil {
			if ne, ok := err.(net.Error); ok && ne.Timeout() {
				return 0, nil, nil, secureHTTPFail("response_header_timeout")
			}
			return 0, nil, nil, secureHTTPFail("connection_failed")
		}
		if n == 0 {
			return 0, nil, nil, secureHTTPFail("connection_failed")
		}
	}
}

type timedConnReader struct {
	conn    net.Conn
	startMS int64
	totalMS int64
	clock   func() int64
}

func (r *timedConnReader) Read(maxBytes int, timeoutMS int) ([]byte, error) {
	now := r.clock()
	if now-r.startMS >= r.totalMS {
		return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
	}
	remaining := r.totalMS - (now - r.startMS)
	deadline := int64(timeoutMS)
	if remaining < deadline {
		deadline = remaining
	}
	if deadline <= 0 {
		return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
	}
	_ = r.conn.SetReadDeadline(time.Now().Add(time.Duration(deadline) * time.Millisecond))
	buf := make([]byte, maxBytes)
	n, err := r.conn.Read(buf)
	if n > 0 {
		return buf[:n], nil
	}
	if err != nil {
		if ne, ok := err.(net.Error); ok && ne.Timeout() {
			if r.clock()-r.startMS >= r.totalMS {
				return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
			}
			return nil, &BoundedIOError{ReasonCode: "read_idle_timeout"}
		}
		if err == io.EOF {
			return []byte{}, nil
		}
		return nil, err
	}
	return []byte{}, nil
}

type pendingConnReader struct {
	timedConnReader
	pending []byte
}

func (r *pendingConnReader) Read(maxBytes int, timeoutMS int) ([]byte, error) {
	if len(r.pending) > 0 {
		if maxBytes > len(r.pending) {
			maxBytes = len(r.pending)
		}
		out := r.pending[:maxBytes]
		r.pending = r.pending[maxBytes:]
		return out, nil
	}
	return r.timedConnReader.Read(maxBytes, timeoutMS)
}

type lengthLimitedReader struct {
	reader    pendingConnReader
	remaining int
}

func (r *lengthLimitedReader) Read(maxBytes int, timeoutMS int) ([]byte, error) {
	if r.remaining <= 0 {
		return []byte{}, nil
	}
	if maxBytes > r.remaining {
		maxBytes = r.remaining
	}
	data, err := r.reader.Read(maxBytes, timeoutMS)
	r.remaining -= len(data)
	return data, err
}

const maxChunkLineBytes = 8192

func isHexDigit(b byte) bool {
	return (b >= '0' && b <= '9') || (b >= 'a' && b <= 'f') || (b >= 'A' && b <= 'F')
}

func readChunkedBody(conn net.Conn, profile BudgetProfile, startMS, totalMS int64, clock func() int64, initial []byte) ([]byte, error) {
	limit := profile.ResponseBodyBytes()
	pending := append([]byte(nil), initial...)
	result := make([]byte, 0, 1024)

	readSome := func(size int) error {
		for len(pending) < size {
			now := clock()
			if now-startMS >= totalMS {
				return &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
			}
			remaining := totalMS - (now - startMS)
			timeoutMS := int64(1000)
			if remaining < timeoutMS {
				timeoutMS = remaining
			}
			if timeoutMS <= 0 {
				return &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
			}
			_ = conn.SetReadDeadline(time.Now().Add(time.Duration(timeoutMS) * time.Millisecond))
			tmp := make([]byte, size-len(pending))
			n, err := conn.Read(tmp)
			if n > 0 {
				pending = append(pending, tmp[:n]...)
				continue
			}
			if err != nil {
				if ne, ok := err.(net.Error); ok && ne.Timeout() {
					if clock()-startMS >= totalMS {
						return &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
					}
					return &BoundedIOError{ReasonCode: "read_idle_timeout"}
				}
				return secureHTTPFail("connection_failed")
			}
			return secureHTTPFail("connection_failed")
		}
		return nil
	}

	readLine := func() ([]byte, error) {
		for {
			if idx := bytes.Index(pending, []byte("\r\n")); idx >= 0 {
				line := pending[:idx]
				pending = pending[idx+2:]
				return line, nil
			}
			if len(pending) >= maxChunkLineBytes {
				return nil, secureHTTPFail("invalid_chunked_encoding")
			}
			now := clock()
			if now-startMS >= totalMS {
				return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
			}
			remaining := totalMS - (now - startMS)
			timeoutMS := int64(1000)
			if remaining < timeoutMS {
				timeoutMS = remaining
			}
			if timeoutMS <= 0 {
				return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
			}
			_ = conn.SetReadDeadline(time.Now().Add(time.Duration(timeoutMS) * time.Millisecond))
			tmp := make([]byte, 4096)
			n, err := conn.Read(tmp)
			if n > 0 {
				pending = append(pending, tmp[:n]...)
				continue
			}
			if err != nil {
				if ne, ok := err.(net.Error); ok && ne.Timeout() {
					if clock()-startMS >= totalMS {
						return nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
					}
					return nil, &BoundedIOError{ReasonCode: "read_idle_timeout"}
				}
				return nil, secureHTTPFail("connection_failed")
			}
			return nil, secureHTTPFail("connection_failed")
		}
	}

	for {
		line, err := readLine()
		if err != nil {
			return nil, err
		}
		if bytes.Contains(line, []byte(";")) {
			return nil, secureHTTPFail("chunk_extension_not_allowed")
		}
		if len(line) == 0 {
			return nil, secureHTTPFail("invalid_chunked_encoding")
		}
		for _, b := range line {
			if !isHexDigit(b) {
				return nil, secureHTTPFail("invalid_chunked_encoding")
			}
		}
		sizeValue, err := strconv.ParseInt(string(line), 16, 64)
		if err != nil {
			return nil, secureHTTPFail("invalid_chunked_encoding")
		}
		if sizeValue == 0 {
			trailerLine, err := readLine()
			if err != nil {
				return nil, err
			}
			if len(trailerLine) != 0 {
				return nil, secureHTTPFail("invalid_chunked_trailer")
			}
			break
		}
		if sizeValue > int64(limit-len(result)) {
			return nil, secureHTTPFail("response_body_too_large")
		}
		if err := readSome(int(sizeValue)); err != nil {
			return nil, err
		}
		result = append(result, pending[:sizeValue]...)
		pending = pending[sizeValue:]
		if err := readSome(2); err != nil {
			return nil, err
		}
		if !bytes.Equal(pending[:2], []byte("\r\n")) {
			return nil, secureHTTPFail("invalid_chunked_encoding")
		}
		pending = pending[2:]
	}
	return result, nil
}

func validateFramingHeaders(headerMap map[string][]string) error {
	teValues := headerMap["transfer-encoding"]
	clValues := headerMap["content-length"]
	if len(teValues) == 0 {
		return nil
	}
	if len(clValues) > 0 {
		return secureHTTPFail("invalid_transfer_encoding")
	}
	codings := make([]string, 0, 4)
	for _, value := range teValues {
		for _, part := range strings.Split(value, ",") {
			codings = append(codings, strings.ToLower(strings.Trim(part, " \t")))
		}
	}
	if len(codings) != 1 || codings[0] != "chunked" {
		return secureHTTPFail("invalid_transfer_encoding")
	}
	return nil
}

// ReadRawResponse reads a raw HTTP/1.1 response without keep-alive reuse.
func ReadRawResponse(conn net.Conn, method string, profile BudgetProfile, startMS, totalMS int64, clock func() int64) (int, []Header, []byte, error) {
	if totalMS <= 0 {
		return 0, nil, nil, &BoundedIOError{ReasonCode: "total_timeout_exceeded"}
	}
	budget := NewTransportBudget()
	headerTimeout := budget.ResponseHeaderTimeoutMS()
	if int64(headerTimeout) > totalMS {
		headerTimeout = int(totalMS)
	}
	status, headers, rest, err := readResponseHeads(conn, headerTimeout)
	if err != nil {
		return 0, nil, nil, err
	}
	headerMap := map[string][]string{}
	for _, header := range headers {
		key := strings.ToLower(header.Name)
		headerMap[key] = append(headerMap[key], header.Value)
	}
	if method == "HEAD" || status == 204 || status == 304 {
		if err := validateFramingHeaders(headerMap); err != nil {
			return 0, nil, nil, err
		}
		return status, headers, []byte{}, nil
	}
	if err := validateFramingHeaders(headerMap); err != nil {
		return 0, nil, nil, err
	}
	if enc := headerMap["content-encoding"]; len(enc) > 0 && strings.ToLower(enc[0]) != "identity" {
		return 0, nil, nil, secureHTTPFail("unsupported_content_encoding")
	}
	if len(headerMap["transfer-encoding"]) > 0 {
		body, err := readChunkedBody(conn, profile, startMS, totalMS, clock, rest)
		if err != nil {
			return 0, nil, nil, err
		}
		return status, headers, body, nil
	}
	clValues := headerMap["content-length"]
	if len(clValues) > 0 {
		anyValues := make([]any, 0, len(clValues))
		for _, value := range clValues {
			anyValues = append(anyValues, value)
		}
		declared, err := ValidateContentLengths(anyValues, profile)
		if err != nil {
			return 0, nil, nil, err
		}
		reader := &lengthLimitedReader{reader: pendingConnReader{timedConnReader: timedConnReader{conn: conn, startMS: startMS, totalMS: totalMS, clock: clock}, pending: rest}, remaining: declared}
		result, err := ReadBoundedWithClock(profile, reader, clock)
		if err != nil {
			return 0, nil, nil, err
		}
		if result.TotalRead < declared {
			return 0, nil, nil, secureHTTPFail("connection_failed")
		}
		return status, headers, result.Data, nil
	}
	reader := &pendingConnReader{timedConnReader: timedConnReader{conn: conn, startMS: startMS, totalMS: totalMS, clock: clock}, pending: rest}
	result, err := ReadBoundedWithClock(profile, reader, clock)
	if err != nil {
		return 0, nil, nil, err
	}
	return status, headers, result.Data, nil
}

File: crawler/security/http_transport.py
"""Isolated secure HTTP/1.1 transport primitives for TASK-022D-3.

This module performs raw HTTP/1.1 framing over a pinned connection. It does
not use requests/httpx, does not read environment proxies, and never follows
redirects itself.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Sequence
from urllib.parse import urlsplit

from crawler.security.bounded_io import (
    BoundedIOError,
    read_bounded,
    validate_content_lengths,
)
from crawler.security.transport_budget import BUDGET, budget_for

_MAX_HEADER_BYTES = BUDGET.response_headers_bytes
_ALLOWED_METHODS = frozenset({"GET", "HEAD", "POST"})
_FORBIDDEN_HEADERS = frozenset(
    {"host", "transfer-encoding", "proxy-authorization", "connection", "content-length"}
)
_TOKEN_CHARS = frozenset("!#$%&'*+-.^_`|~0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz")
_CHUNK_LINE_MAX = 8192


class SecureHTTPError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


@dataclass(frozen=True, slots=True)
class SecureHTTPRequest:
    profile: str
    method: str
    url: str
    headers: tuple[tuple[str, str], ...] = ()
    body: bytes = b""


@dataclass(frozen=True, slots=True)
class SecureHTTPResponse:
    status_code: int
    final_url: str
    headers: tuple[tuple[str, str], ...]
    body: bytes
    redirect_hops: int


def _invalid_header(value: str) -> bool:
    return any(ch in value for ch in "\r\n\x00")


def _is_http_token(value: str) -> bool:
    return bool(value) and all(ch in _TOKEN_CHARS for ch in value)


def _trim_ows(value: bytes) -> bytes:
    return value.strip(b" \t")


def validate_request(request: SecureHTTPRequest) -> None:
    if request.method not in _ALLOWED_METHODS:
        raise SecureHTTPError("method_not_allowed")
    if not isinstance(request.body, bytes):
        raise SecureHTTPError("invalid_request_body")
    if request.method in ("GET", "HEAD") and request.body:
        raise SecureHTTPError("method_body_not_allowed")
    if request.body:
        from crawler.security.bounded_io import check_request_body_size

        check_request_body_size(len(request.body))
    for name, value in request.headers:
        if _invalid_header(name) or _invalid_header(value):
            raise SecureHTTPError("invalid_header")
        if not _is_http_token(name):
            raise SecureHTTPError("invalid_header")
        if name.lower() in _FORBIDDEN_HEADERS:
            raise SecureHTTPError("forbidden_header")


def build_request_bytes(request: SecureHTTPRequest, host_header: str) -> bytes:
    validate_request(request)
    parsed = urlsplit(request.url)
    path = parsed.path or "/"
    if parsed.query:
        path += "?" + parsed.query
    lines = [f"{request.method} {path} HTTP/1.1", f"Host: {host_header}"]
    sent_encoding = False
    for name, value in request.headers:
        if name.lower() == "accept-encoding":
            sent_encoding = True
        lines.append(f"{name}: {value}")
    if not sent_encoding:
        lines.append("Accept-Encoding: identity")
    if request.method == "POST":
        lines.append(f"Content-Length: {len(request.body)}")
    lines.append("Connection: close")
    header_block = "\r\n".join(lines) + "\r\n\r\n"
    return header_block.encode("latin-1") + request.body


def _parse_status_and_headers(head: bytes) -> tuple[int, list[tuple[str, str]]]:
    lines = head.split(b"\r\n")
    first_line = lines[0]
    for byte in first_line:
        if byte < 0x20 and byte not in (0x09, 0x20):
            raise SecureHTTPError("invalid_http_response")
        if byte == 0x7f:
            raise SecureHTTPError("invalid_http_response")
    status_parts = first_line.split(b" ", 2)
    if len(status_parts) < 2:
        raise SecureHTTPError("invalid_http_response")
    if status_parts[0] not in (b"HTTP/1.0", b"HTTP/1.1"):
        raise SecureHTTPError("invalid_http_response")
    status_text = status_parts[1]
    if len(status_text) != 3 or not status_text.isdigit():
        raise SecureHTTPError("invalid_http_response")
    status = int(status_text)
    headers: list[tuple[str, str]] = []
    for line in lines[1:]:
        if not line:
            raise SecureHTTPError("invalid_http_response")
        if line.startswith((b" ", b"\t")):
            raise SecureHTTPError("invalid_http_response")
        if b"\r" in line or b"\n" in line or b"\x00" in line:
            raise SecureHTTPError("invalid_http_response")
        name, sep, value = line.partition(b":")
        if not sep or not name:
            raise SecureHTTPError("invalid_http_response")
        name_text = name.decode("latin-1")
        if not _is_http_token(name_text):
            raise SecureHTTPError("invalid_http_response")
        value_text = _trim_ows(value).decode("latin-1")
        if "\x00" in value_text or "\r" in value_text or "\n" in value_text:
            raise SecureHTTPError("invalid_http_response")
        headers.append((name_text, value_text))
    return status, headers


def _read_response_headers(conn: Any, timeout_ms: int) -> tuple[int, list[tuple[str, str]], bytes]:
    conn.settimeout(timeout_ms / 1000.0)
    pending = bytearray()
    used = 0
    try:
        while True:
            delimiter = pending.find(b"\r\n\r\n")
            if delimiter >= 0:
                head_end = delimiter + 4
                total = used + head_end
                if total > _MAX_HEADER_BYTES:
                    raise SecureHTTPError("response_headers_too_large")
                status, headers = _parse_status_and_headers(bytes(pending[:delimiter]))
                del pending[:head_end]
                used = total
                if status == 101:
                    raise SecureHTTPError("protocol_upgrade_not_allowed")
                if 100 <= status <= 199:
                    continue
                if not (200 <= status <= 599):
                    raise SecureHTTPError("invalid_http_response")
                return status, headers, bytes(pending)
            if used + len(pending) >= _MAX_HEADER_BYTES:
                raise SecureHTTPError("response_headers_too_large")
            chunk = conn.recv(4096)
            if not chunk:
                raise SecureHTTPError("connection_failed")
            pending.extend(chunk)
    except TimeoutError as exc:
        raise SecureHTTPError("response_header_timeout") from exc
    raise SecureHTTPError("connection_failed")


class _SocketTimedReader:
    deadline_capable = True

    def __init__(self, conn: Any, start_ms: int, total_ms: int, clock, pending: bytes = b""):
        self.conn = conn
        self.start_ms = start_ms
        self.total_ms = total_ms
        self.clock = clock
        self.pending = pending

    def read(self, max_bytes: int, timeout_ms: int) -> bytes:
        if self.pending:
            out = self.pending[:max_bytes]
            self.pending = self.pending[max_bytes:]
            return out
        now = self.clock()
        if now - self.start_ms >= self.total_ms:
            raise BoundedIOError("total_timeout_exceeded")
        remaining = self.total_ms - (now - self.start_ms)
        read_timeout = min(timeout_ms, remaining)
        self.conn.settimeout(max(1, read_timeout) / 1000.0)
        try:
            data = self.conn.recv(max_bytes)
        except TimeoutError:
            now2 = self.clock()
            if now2 - self.start_ms >= self.total_ms:
                raise BoundedIOError("total_timeout_exceeded")
            raise TimeoutError()
        if not data:
            return b""
        return data


class _LengthLimitedReader(_SocketTimedReader):
    def __init__(self, conn, start_ms, total_ms, clock, limit, pending=b""):
        super().__init__(conn, start_ms, total_ms, clock, pending)
        self.remaining = limit

    def read(self, max_bytes: int, timeout_ms: int) -> bytes:
        if self.remaining <= 0:
            return b""
        allowed = min(max_bytes, self.remaining)
        data = super().read(allowed, timeout_ms)
        self.remaining -= len(data)
        return data


class _ClockAdapter:
    def __init__(self, fn):
        self.fn = fn

    def monotonic_ms(self):
        return self.fn()


class _ChunkedTimedReader(_SocketTimedReader):
    def __init__(self, conn, start_ms, total_ms, clock, pending=b""):
        super().__init__(conn, start_ms, total_ms, clock, pending)
        self._buffer = bytearray(pending)
        self._done = False

    def _read_line(self):
        data = bytearray()
        while True:
            idx = self._buffer.find(b"\r\n")
            if idx >= 0:
                line = bytes(self._buffer[:idx])
                del self._buffer[:idx + 2]
                return line
            chunk = super().read(4096, 1000)
            if not chunk:
                raise SecureHTTPError("connection_failed")
            self._buffer.extend(chunk)

    def _read_exact(self, size):
        while len(self._buffer) < size:
            chunk = super().read(size - len(self._buffer), 1000)
            if not chunk:
                raise SecureHTTPError("connection_failed")
            self._buffer.extend(chunk)
        out = bytes(self._buffer[:size])
        del self._buffer[:size]
        return out

    def read(self, max_bytes, timeout_ms):
        while not self._done and len(self._buffer) < max_bytes:
            line = self._read_line()
            size_text = line.split(b";", 1)[0].strip()
            try:
                size = int(size_text, 16)
            except ValueError:
                raise SecureHTTPError("invalid_chunked_encoding")
            if size == 0:
                self._read_exact(2)
                self._done = True
                break
            data = self._read_exact(size)
            trailer = self._read_exact(2)
            if trailer != b"\r\n":
                raise SecureHTTPError("invalid_chunked_encoding")
            self._buffer.extend(data)
        out = bytes(self._buffer[:max_bytes])
        del self._buffer[:max_bytes]
        return out


def _read_chunked_body_manual(conn, profile, start_ms, total_ms, clock, initial=b""):
    selected = budget_for(profile)
    pending = bytearray(initial)
    result = bytearray()

    def fill_needed(size):
        while len(pending) < size:
            now = clock()
            if now - start_ms >= total_ms:
                raise SecureHTTPError("total_timeout_exceeded")
            conn.settimeout(min(1000, total_ms - (now - start_ms)) / 1000.0)
            try:
                data = conn.recv(size - len(pending))
            except TimeoutError as exc:
                if now - start_ms >= total_ms:
                    raise SecureHTTPError("total_timeout_exceeded") from exc
                raise SecureHTTPError("read_idle_timeout") from exc
            if not data:
                raise SecureHTTPError("connection_failed")
            pending.extend(data)

    def read_line(max_len=_CHUNK_LINE_MAX):
        while True:
            idx = pending.find(b"\r\n")
            if idx >= 0:
                line = bytes(pending[:idx])
                del pending[:idx + 2]
                return line
            if len(pending) >= max_len:
                raise SecureHTTPError("invalid_chunked_encoding")
            now = clock()
            if now - start_ms >= total_ms:
                raise SecureHTTPError("total_timeout_exceeded")
            conn.settimeout(min(1000, total_ms - (now - start_ms)) / 1000.0)
            try:
                data = conn.recv(4096)
            except TimeoutError as exc:
                if now - start_ms >= total_ms:
                    raise SecureHTTPError("total_timeout_exceeded") from exc
                raise SecureHTTPError("read_idle_timeout") from exc
            if not data:
                raise SecureHTTPError("connection_failed")
            pending.extend(data)

    while True:
        line = read_line()
        if b";" in line:
            raise SecureHTTPError("chunk_extension_not_allowed")
        if not line or any(ch not in b"0123456789abcdefABCDEF" for ch in line):
            raise SecureHTTPError("invalid_chunked_encoding")
        size = int(line, 16)
        if size > (1 << 63) - 1:
            raise SecureHTTPError("invalid_chunked_encoding")
        if size == 0:
            if read_line():
                raise SecureHTTPError("invalid_chunked_trailer")
            break
        fill_needed(size)
        result.extend(pending[:size])
        del pending[:size]
        if len(result) > selected.response_body_bytes:
            raise SecureHTTPError("response_body_too_large")
        fill_needed(2)
        if bytes(pending[:2]) != b"\r\n":
            raise SecureHTTPError("invalid_chunked_encoding")
        del pending[:2]
    return bytes(result)


def _validate_framing_headers(header_map: dict[str, list[str]]) -> None:
    te_values = header_map.get("transfer-encoding")
    cl_values = header_map.get("content-length")
    if te_values is None:
        return
    if cl_values is not None:
        raise SecureHTTPError("invalid_transfer_encoding")
    codings: list[str] = []
    for value in te_values:
        for part in value.split(","):
            coding = part.strip(" \t").lower()
            codings.append(coding)
    if codings != ["chunked"]:
        raise SecureHTTPError("invalid_transfer_encoding")


def _read_body(
    conn,
    request_method: str,
    status: int,
    headers: Sequence[tuple[str, str]],
    profile: str,
    start_ms: int,
    total_ms: int,
    clock,
    initial: bytes = b"",
) -> bytes:
    header_map: dict[str, list[str]] = {}
    for name, value in headers:
        header_map.setdefault(name.lower(), []).append(value)
    if request_method == "HEAD" or status in (204, 304):
        _validate_framing_headers(header_map)
        return b""
    _validate_framing_headers(header_map)
    encoding = header_map.get("content-encoding", [])
    if encoding and encoding[0].lower() not in ("identity", ""):
        raise SecureHTTPError("unsupported_content_encoding")
    if header_map.get("transfer-encoding") is not None:
        return _read_chunked_body_manual(conn, profile, start_ms, total_ms, clock, initial)
    length_values = header_map.get("content-length")
    if length_values is not None:
        declared = validate_content_lengths([v for v in length_values], profile)
        if declared is None:
            declared = 0
        reader = _LengthLimitedReader(conn, start_ms, total_ms, clock, declared, initial)
        result = read_bounded(profile, reader, clock=_ClockAdapter(clock))
        if result.total_read < declared:
            raise SecureHTTPError("connection_failed")
        return result.data
    reader = _SocketTimedReader(conn, start_ms, total_ms, clock, initial)
    result = read_bounded(profile, reader, clock=_ClockAdapter(clock))
    return result.data


def read_raw_response(
    conn,
    request_method: str,
    profile: str,
    start_ms: int,
    total_ms: int,
    clock,
) -> tuple[int, tuple[tuple[str, str], ...], bytes]:
    if total_ms <= 0:
        raise SecureHTTPError("total_timeout_exceeded")
    budget_for(profile)
    header_timeout = min(BUDGET.response_header_timeout_ms, total_ms)
    status, headers, rest = _read_response_headers(conn, header_timeout)
    body = _read_body(
        conn,
        request_method,
        status,
        headers,
        profile,
        start_ms,
        total_ms,
        clock,
        rest,
    )
    return status, tuple(headers), body
